File: 1392-longest-happy-prefix/1392-longest-happy-prefix.py
class Solution:
    def longestPrefix(self, s: str) -> str:
        n = len(s)
        dp = [0]*n
        for i in range(1,n):
            j = dp[i-1]
            while j>0 and s[j]!=s[i]:
                j = dp[j-1]
            dp[i] = j + (1 if s[i]==s[j] else 0)
        # print(dp)
        l = dp[n-1]
        return s[:l] if l>0 else ""

File: 1392-longest-happy-prefix/1392-longest-happy-prefix.java
class Solution {
    public String longestPrefix(String s) {
        int n = s.length();
        int[] dp = new int[n];
        
        dp[0] = 0;
        for (int i = 1; i<n; i++) {
            int j = dp[i-1];
            while (j>0 && s.charAt(j)!=s.charAt(i)) 
                j = dp[j-1];
            dp[i] = j + (s.charAt(j)==s.charAt(i) ? 1:0);
        }
        return s.substring(0, dp[n-1]);
    }
}